File: lib/data/repositories/football_repository.dart
import '../../core/constants/league_constants.dart';
import '../models/fixture.dart';
import '../models/fixture_details.dart';
import '../models/league_summary.dart';
import '../models/standing_entry.dart';
import '../models/team.dart';
import '../services/football_api_client.dart';

class HomeFeed {
  final List<Fixture> live;
  final List<Fixture> upcoming;
  final List<Fixture> todayResults;
  final List<Fixture> yesterdayResults;
  const HomeFeed({
    required this.live,
    required this.upcoming,
    required this.todayResults,
    required this.yesterdayResults,
  });

  bool get isEmpty =>
      live.isEmpty && upcoming.isEmpty && todayResults.isEmpty && yesterdayResults.isEmpty;

  List<Fixture> get all => [...live, ...upcoming, ...todayResults, ...yesterdayResults];
}

class RoundNavigation {
  final List<String> rounds;
  final int currentIndex;
  const RoundNavigation({required this.rounds, required this.currentIndex});

  String? get current =>
      currentIndex >= 0 && currentIndex < rounds.length ? rounds[currentIndex] : null;
  bool get hasPrevious => currentIndex > 0;
  bool get hasNext => currentIndex >= 0 && currentIndex < rounds.length - 1;
}

class FootballRepository {
  final FootballApiClient _client;
  FootballRepository({FootballApiClient? client}) : _client = client ?? FootballApiClient.instance;

  /// v1 hardcoded season=2022, which goes stale every year. European
  /// domestic leagues run Aug-May, so before August we're still in the
  /// season that *started* the previous calendar year.
  int currentSeason([DateTime? now]) {
    final n = now ?? DateTime.now();
    return n.month >= 8 ? n.year : n.year - 1;
  }

  String _fmt(DateTime d) =>
      "${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}";

  /// Raw candidate pool for a given day, filtered to leagues the app's
  /// API-Sports plan actually covers ([trackedLeagueIds]). This is the
  /// pool the AI curation service later ranks/selects from — see
  /// AiCurationService in data/services/ai.
  Future<List<Fixture>> fixturesForDate(DateTime date) async {
    final raw = await _client.fixturesByDate(_fmt(date));
    return raw
        .where((f) => trackedLeagueIds.contains(f["league"]["id"]))
        .map((f) => Fixture.fromJson(f as Map<String, dynamic>))
        .toList();
  }

  Future<HomeFeed> homeFeed({DateTime? date}) async {
    final today = date ?? DateTime.now();
    final yesterday = today.subtract(const Duration(days: 1));

    final todayFixtures = await fixturesForDate(today);
    final yesterdayFixtures = await fixturesForDate(yesterday);

    return HomeFeed(
      live: todayFixtures.where((f) => f.isLive).toList(),
      upcoming: todayFixtures.where((f) => f.isScheduled).toList(),
      todayResults: todayFixtures.where((f) => f.isFinished).toList(),
      yesterdayResults: yesterdayFixtures.where((f) => f.isFinished).toList(),
    );
  }

  Future<List<Fixture>> leagueFixtures({required int leagueId, String? round}) async {
    final season = currentSeason();
    final r = round ?? await _client.currentRound(leagueId: leagueId, season: season);
    if (r == null) return [];
    final raw = await _client.fixturesByLeagueAndRound(
      leagueId: leagueId,
      season: season,
      round: r,
    );
    return raw.map((f) => Fixture.fromJson(f as Map<String, dynamic>)).toList();
  }

  /// Rounds list + index of the current round, so the Fixtures page can
  /// move Previous round / Current round / Next round (section 7).
  Future<RoundNavigation> roundNavigation({required int leagueId}) async {
    final season = currentSeason();
    final rounds = await _client.allRounds(leagueId: leagueId, season: season);
    final current = await _client.currentRound(leagueId: leagueId, season: season);
    final idx = current != null ? rounds.indexOf(current) : -1;
    return RoundNavigation(rounds: rounds, currentIndex: idx);
  }

  /// The list of team standing rows for the league's main group/table.
  /// (API-Football nests this as `response[0].league.standings[0]` — a
  /// *list* of row objects, not a map; [FootballApiClient.standings]
  /// already unwraps down to that inner list.)
  Future<List<dynamic>?> leagueStandingsRaw({required int leagueId}) =>
      _client.standings(leagueId: leagueId, season: currentSeason());

  Future<List<StandingEntry>> leagueStandings({required int leagueId}) async {
    final raw = await leagueStandingsRaw(leagueId: leagueId);
    if (raw == null) return [];
    return raw.map((e) => StandingEntry.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<TopPlayerEntry>> leagueTopScorers({required int leagueId}) async {
    final raw = await _client.topScorers(leagueId: leagueId, season: currentSeason());
    return raw.map((e) => TopPlayerEntry.fromJson(e as Map<String, dynamic>, assists: false)).toList();
  }

  Future<List<TopPlayerEntry>> leagueTopAssists({required int leagueId}) async {
    final raw = await _client.topAssists(leagueId: leagueId, season: currentSeason());
    return raw.map((e) => TopPlayerEntry.fromJson(e as Map<String, dynamic>, assists: true)).toList();
  }

  /// Looks up a single fixture by id — used when Fixture Details is opened
  /// without an already-fetched [Fixture] object at hand (e.g. deep link).
  Future<Fixture?> fixtureById(int fixtureId) async {
    final raw = await _client.fixtureById(fixtureId);
    return raw != null ? Fixture.fromJson(raw) : null;
  }

  Future<List<Team>> searchTeams(String query) async {
    final raw = await _client.searchTeams(query);
    return raw.map((t) => Team.fromJson(t as Map<String, dynamic>)).toList();
  }

  Future<List<LeagueSummary>> searchLeagues(String query) async {
    final raw = await _client.searchLeagues(query);
    return raw.map((l) => LeagueSummary.fromJson(l as Map<String, dynamic>)).toList();
  }

  /// Everything the Fixture Details screen needs, fetched in parallel.
  /// Each piece degrades gracefully to empty/null if API-Football simply
  /// doesn't have it yet for this fixture (e.g. events before kickoff).
  Future<FixtureDetails> fixtureDetails(Fixture fixture) async {
    final results = await Future.wait([
      _client.fixtureEvents(fixture.id),
      _client.fixtureStatistics(fixture.id),
      _client.fixtureLineups(fixture.id),
      _client.headToHead(team1Id: fixture.homeTeamId, team2Id: fixture.awayTeamId),
    ]);

    return FixtureDetails(
      fixture: fixture,
      events: (results[0] as List<dynamic>)
          .map((e) => FixtureEvent.fromJson(e as Map<String, dynamic>))
          .toList(),
      statistics: (results[1] as List<dynamic>)
          .map((s) => TeamStatistics.fromJson(s as Map<String, dynamic>))
          .toList(),
      lineups: (results[2] as List<dynamic>)
          .map((l) => TeamLineup.fromJson(l as Map<String, dynamic>))
          .toList(),
      headToHead: (results[3] as List<dynamic>)
          .map((f) => Fixture.fromJson(f as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: frontend/src/services/footballRepository.ts
import { footballApi } from "./footballApi";
import { trackedLeagueIds } from "@/utils/leagueConstants";
import {
  type Fixture,
  type FixtureDetails,
  type LeagueSummary,
  type StandingEntry,
  type Team,
  type TopPlayerEntry,
  fixtureStatusGroup,
  parseFixture,
  parseFixtureEvent,
  parseLeagueSummary,
  parseStandingEntry,
  parseTeam,
  parseTeamLineup,
  parseTeamStatistics,
  parseTopPlayerEntry,
} from "@/types/models";

export interface HomeFeed {
  live: Fixture[];
  upcoming: Fixture[];
  todayResults: Fixture[];
  yesterdayResults: Fixture[];
}

export function homeFeedAll(feed: HomeFeed): Fixture[] {
  return [...feed.live, ...feed.upcoming, ...feed.todayResults, ...feed.yesterdayResults];
}

export function homeFeedIsEmpty(feed: HomeFeed): boolean {
  return feed.live.length === 0 && feed.upcoming.length === 0 && feed.todayResults.length === 0 && feed.yesterdayResults.length === 0;
}

export interface RoundNavigation {
  rounds: string[];
  currentIndex: number;
}

/** European domestic leagues run Aug-May; before August we're still in the season that started the previous year. */
export function currentSeason(now: Date = new Date()): number {
  return now.getMonth() + 1 >= 8 ? now.getFullYear() : now.getFullYear() - 1;
}

function fmtDate(d: Date): string {
  const y = String(d.getFullYear()).padStart(4, "0");
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

export const footballRepository = {
  async fixturesForDate(date: Date): Promise<Fixture[]> {
    const raw = await footballApi.fixturesByDate(fmtDate(date));
    return raw.filter((f) => trackedLeagueIds.includes(f.league.id)).map(parseFixture);
  },

  async homeFeed(date: Date = new Date()): Promise<HomeFeed> {
    const yesterday = new Date(date);
    yesterday.setDate(yesterday.getDate() - 1);

    const [todayFixtures, yesterdayFixtures] = await Promise.all([
      footballRepository.fixturesForDate(date),
      footballRepository.fixturesForDate(yesterday),
    ]);

    return {
      live: todayFixtures.filter((f) => fixtureStatusGroup(f) === "live"),
      upcoming: todayFixtures.filter((f) => fixtureStatusGroup(f) === "scheduled"),
      todayResults: todayFixtures.filter((f) => fixtureStatusGroup(f) === "finished"),
      yesterdayResults: yesterdayFixtures.filter((f) => fixtureStatusGroup(f) === "finished"),
    };
  },

  async leagueFixtures(leagueId: number, round?: string | null): Promise<Fixture[]> {
    const season = currentSeason();
    const r = round ?? (await footballApi.currentRound(leagueId, season));
    if (!r) return [];
    const raw = await footballApi.fixturesByLeagueAndRound(leagueId, season, r);
    return raw.map(parseFixture);
  },

  async roundNavigation(leagueId: number): Promise<RoundNavigation> {
    const season = currentSeason();
    const [rounds, current] = await Promise.all([
      footballApi.allRounds(leagueId, season),
      footballApi.currentRound(leagueId, season),
    ]);
    const idx = current != null ? rounds.indexOf(current) : -1;
    return { rounds, currentIndex: idx };
  },

  async leagueStandings(leagueId: number): Promise<StandingEntry[]> {
    const raw = await footballApi.standings(leagueId, currentSeason());
    if (!raw) return [];
    return raw.map(parseStandingEntry);
  },

  async leagueTopScorers(leagueId: number): Promise<TopPlayerEntry[]> {
    const raw = await footballApi.topScorers(leagueId, currentSeason());
    return raw.map((e) => parseTopPlayerEntry(e, false));
  },

  async leagueTopAssists(leagueId: number): Promise<TopPlayerEntry[]> {
    const raw = await footballApi.topAssists(leagueId, currentSeason());
    return raw.map((e) => parseTopPlayerEntry(e, true));
  },

  async fixtureById(fixtureId: number): Promise<Fixture | null> {
    const raw = await footballApi.fixtureById(fixtureId);
    return raw ? parseFixture(raw) : null;
  },

  async searchTeams(query: string): Promise<Team[]> {
    const raw = await footballApi.searchTeams(query);
    return raw.map(parseTeam);
  },

  async searchLeagues(query: string): Promise<LeagueSummary[]> {
    const raw = await footballApi.searchLeagues(query);
    return raw.map(parseLeagueSummary);
  },

  async fixtureDetails(fixture: Fixture): Promise<FixtureDetails> {
    const [events, statistics, lineups, headToHead] = await Promise.all([
      footballApi.fixtureEvents(fixture.id),
      footballApi.fixtureStatistics(fixture.id),
      footballApi.fixtureLineups(fixture.id),
      footballApi.headToHead(fixture.homeTeamId, fixture.awayTeamId),
    ]);

    return {
      fixture,
      events: events.map(parseFixtureEvent),
      statistics: statistics.map(parseTeamStatistics),
      lineups: lineups.map(parseTeamLineup),
      headToHead: headToHead.map(parseFixture),
    };
  },
};
